chore(Analysis/Normed/Lp/WithLp): rename `WithLp.equiv`, make arguments implicit and change the direction of the equivalence

Paul-Lez opened the pull request:
---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

YaelDillies commented on Mathlib/Analysis/Normed/Lp/ProdLp.lean:
```diff
@@ -458,6 +515,14 @@ theorem prod_continuous_equiv : Continuous (WithLp.equiv p (α × β)) :=
+  continuous_id
+
+@[continuity]
+theorem prod_continuous_ofLp : Continuous (@WithLp.ofLp p (α × β)).symm :=
+  continuous_id
```
`ofLp.symm` should be simplified to `toLp`, so no lemma should be written about it. Or maybe you meant
```suggestion
theorem prod_continuous_ofLp : Continuous (@WithLp.ofLp p (α × β)) :=
  continuous_id
```
?

Paul-Lez replied:
Oh yeah this was a typo!

YaelDillies commented on Mathlib/Analysis/Normed/Lp/ProdLp.lean:
```diff
@@ -804,16 +972,20 @@ def idemFst : AddMonoid.End (WithLp p (α × β)) := (AddMonoidHom.inl α β).co
+lemma idemFst_apply (x : WithLp p (α × β)) : idemFst x = WithLp.toLp (x.1, 0) := rfl
 
-lemma idemSnd_apply (x : WithLp p (α × β)) : idemSnd x = (WithLp.equiv _ _).symm (0, x.2) := rfl
+lemma idemSnd_apply (x : WithLp p (α × β)) : idemSnd x =  WithLp.toLp (0, x.2) := rfl
```
```suggestion
lemma idemSnd_apply (x : WithLp p (α × β)) : idemSnd x = WithLp.toLp (0, x.2) := rfl
```

YaelDillies commented on Mathlib/Analysis/Normed/Lp/WithLp.lean:
```diff
@@ -50,8 +51,19 @@ variable (p : ℝ≥0∞) (K : Type uK) (K' : Type uK') (V : Type uV)
+variable {p V} in
+/-- The canonical equivalence between `V` and `WithLp p V`. This should always be used to convert
+back and forth between the representations. -/
+def toLp : V ≃ WithLp p V := Equiv.refl _
```
nit:
```suggestion
def toLp : V ≃ WithLp p V := .refl _
```
Same below

YaelDillies replied:
It's new in the sense that this syntax didn't exist in Lean 3, but Lean 4 by now isn't so new 😅

Paul-Lez replied:
I meant the style convention of *prefering* `.refl` to `Equiv.refl` :)

YaelDillies commented on Mathlib/Analysis/Normed/Lp/WithLp.lean:
```diff
@@ -85,7 +97,65 @@ instance instModuleFinite
 
-/-! `WithLp.equiv` preserves the module structure. -/
+@[simp]
+theorem toLp_symm_eq : (@toLp p V).symm = toLp (p:= p) := rfl
```
```suggestion
theorem toLp_symm_eq : (@toLp p V).symm = toLp (p := p) := rfl
```
Same below

YaelDillies commented on Mathlib/Analysis/Normed/Lp/WithLp.lean:
```diff
@@ -85,7 +97,65 @@ instance instModuleFinite
+
+/-! `WithLp.toLp` preserves the module structure. -/
+
+@[simp]
+theorem toLp_zero [AddCommGroup V] : WithLp.toLp 0 = (0 : WithLp p V) :=
+  rfl
```
```suggestion
section AddCommGroup
variable [AddCommGroup V]

@[simp] lemma toLp_zero : toLp 0 = (0 : WithLp p V) := rfl
@[simp] lemma ofLp_zero : ofLp (0 : WithLp p V) = 0 := rfl
```
and similarly for other pairs makes it clearer you've got them all

YaelDillies commented on Mathlib/Analysis/Normed/Lp/WithLp.lean:
```diff
@@ -136,6 +206,13 @@ theorem equiv_symm_smul [SMul K V] (c : K) (x' : V) :
 
+/-- `WithLp.toLp` as a linear equivalence. -/
+@[simps -fullyApplied]
+protected def toLpLinear [Semiring K] [AddCommGroup V] [Module K V] : V ≃ₗ[K] WithLp p V :=
```
```suggestion
protected def toLpLinearEquiv [Semiring K] [AddCommGroup V] [Module K V] : V ≃ₗ[K] WithLp p V :=
```

YaelDillies commented on Mathlib/Analysis/Normed/Lp/WithLp.lean:
```diff
@@ -136,6 +206,13 @@ theorem equiv_symm_smul [SMul K V] (c : K) (x' : V) :
 
+/-- `WithLp.toLp` as a linear equivalence. -/
+@[simps -fullyApplied]
+protected def toLpLinear [Semiring K] [AddCommGroup V] [Module K V] : V ≃ₗ[K] WithLp p V :=
+  { LinearEquiv.refl _ _ with
+    toFun := WithLp.toLp
+    invFun := WithLp.ofLp }
```
Can you use `where`?

YaelDillies replied:
You need spread notation:
```
protected def toLpLinear [Semiring K] [AddCommGroup V] [Module K V] : V ≃ₗ[K] WithLp p V where
  __ := LinearEquiv.refl _ _
  toFun := WithLp.toLp
  invFun := WithLp.ofLp
```

YaelDillies commented on Mathlib/Analysis/Normed/Lp/ProdLp.lean:
```diff
@@ -423,11 +429,11 @@ theorem prod_antilipschitzWith_equiv_aux [PseudoEMetricSpace α] [PseudoEMetricS
       (prod_lipschitzWith_equiv_aux p α β).uniformContinuous
   have : (fun x : WithLp p (α × β) × WithLp p (α × β) =>
-    ((WithLp.equiv p (α × β)) x.fst, (WithLp.equiv p (α × β)) x.snd)) = id := by
+    (WithLp.ofLp x.fst, WithLp.ofLp x.snd)) = id := by
     ext i <;> rfl
```
While you're touching this, I believe the following works
```suggestion
    (WithLp.ofLp x.fst, WithLp.ofLp x.snd)) = id := rfl
```

YaelDillies commented on Mathlib/Analysis/Normed/Lp/ProdLp.lean:
```diff
@@ -423,11 +429,11 @@ theorem prod_antilipschitzWith_equiv_aux [PseudoEMetricSpace α] [PseudoEMetricS
 theorem prod_aux_uniformity_eq [PseudoEMetricSpace α] [PseudoEMetricSpace β] :
     𝓤 (WithLp p (α × β)) = 𝓤[instUniformSpaceProd] := by
-  have A : IsUniformInducing (WithLp.equiv p (α × β)) :=
+  have A : IsUniformInducing (@WithLp.ofLp p (α × β)) :=
     (prod_antilipschitzWith_equiv_aux p α β).isUniformInducing
       (prod_lipschitzWith_equiv_aux p α β).uniformContinuous
```
While you're touching this, sounds like this should be a lemma

YaelDillies commented on Mathlib/Analysis/Normed/Lp/ProdLp.lean:
```diff
@@ -439,6 +445,47 @@ theorem prod_aux_cobounded_eq [PseudoMetricSpace α] [PseudoMetricSpace β] :
+  intro x y
+  rcases p.dichotomy with (rfl | h)
+  · simp [edist]
+  · have cancel : p.toReal * (1 / p.toReal) = 1 := mul_div_cancel₀ 1 (zero_lt_one.trans_le h).ne'
+    rw [prod_edist_eq_add (zero_lt_one.trans_le h)]
+    simp only [edist, forall_prop_of_true, one_mul, ENNReal.coe_one, sup_le_iff]
+    constructor
+    · calc
+        edist x.fst y.fst ≤ (edist x.fst y.fst ^ p.toReal) ^ (1 / p.toReal) := by
+          simp only [← ENNReal.rpow_mul, cancel, ENNReal.rpow_one, le_refl]
+        _ ≤ (edist x.fst y.fst ^ p.toReal + edist x.snd y.snd ^ p.toReal) ^ (1 / p.toReal) := by
+          gcongr
+          simp only [self_le_add_right]
+    · calc
+        edist x.snd y.snd ≤ (edist x.snd y.snd ^ p.toReal) ^ (1 / p.toReal) := by
+          simp only [← ENNReal.rpow_mul, cancel, ENNReal.rpow_one, le_refl]
+        _ ≤ (edist x.fst y.fst ^ p.toReal + edist x.snd y.snd ^ p.toReal) ^ (1 / p.toReal) := by
+          gcongr
+          simp only [self_le_add_left]
```
I feel like there's no need to indent so much
```suggestion
  have cancel : p.toReal * (1 / p.toReal) = 1 := mul_div_cancel₀ 1 (zero_lt_one.trans_le h).ne'
  rw [prod_edist_eq_add (zero_lt_one.trans_le h)]
  simp only [edist, forall_prop_of_true, one_mul, ENNReal.coe_one, sup_le_iff]
  constructor
  · calc
      edist x.fst y.fst ≤ (edist x.fst y.fst ^ p.toReal) ^ (1 / p.toReal) := by
        simp only [← ENNReal.rpow_mul, cancel, ENNReal.rpow_one, le_refl]
      _ ≤ (edist x.fst y.fst ^ p.toReal + edist x.snd y.snd ^ p.toReal) ^ (1 / p.toReal) := by
        gcongr
        simp only [self_le_add_right]
  · calc
      edist x.snd y.snd ≤ (edist x.snd y.snd ^ p.toReal) ^ (1 / p.toReal) := by
        simp only [← ENNReal.rpow_mul, cancel, ENNReal.rpow_one, le_refl]
      _ ≤ (edist x.fst y.fst ^ p.toReal + edist x.snd y.snd ^ p.toReal) ^ (1 / p.toReal) := by
        gcongr
        simp only [self_le_add_left]
```

Paul-Lez replied:
Fair - admittedly I just copied and pasted it from the `equiv` version

YaelDillies commented on Mathlib/Analysis/Normed/Lp/ProdLp.lean:
```diff
@@ -439,6 +445,47 @@ theorem prod_aux_cobounded_eq [PseudoMetricSpace α] [PseudoMetricSpace β] :
+          gcongr
+          simp only [self_le_add_left]
+
+theorem prod_antilipschitzWith_ofLp_aux [PseudoEMetricSpace α] [PseudoEMetricSpace β] :
+    AntilipschitzWith ((2 : ℝ≥0) ^ (1 / p).toReal) (@WithLp.ofLp p (α × β)) := by
+  intro x y
+  rcases p.dichotomy with (rfl | h)
+  · simp [edist]
+  · have pos : 0 < p.toReal := by positivity
+    have nonneg : 0 ≤ 1 / p.toReal := by positivity
+    have cancel : p.toReal * (1 / p.toReal) = 1 := mul_div_cancel₀ 1 (ne_of_gt pos)
+    rw [prod_edist_eq_add pos, ENNReal.toReal_div 1 p]
+    simp only [edist, ← one_div, ENNReal.toReal_one]
+    calc
+      (edist x.fst y.fst ^ p.toReal + edist x.snd y.snd ^ p.toReal) ^ (1 / p.toReal) ≤
+          (edist (WithLp.equiv p _ x) (WithLp.equiv p _ y) ^ p.toReal +
+          edist (WithLp.equiv p _ x) (WithLp.equiv p _ y) ^ p.toReal) ^ (1 / p.toReal) := by
+        gcongr <;> simp [edist]
+      _ = (2 ^ (1 / p.toReal) : ℝ≥0) * edist (WithLp.equiv p _ x) (WithLp.equiv p _ y) := by
+        simp only [← two_mul, ENNReal.mul_rpow_of_nonneg _ _ nonneg, ← ENNReal.rpow_mul, cancel,
+          ENNReal.rpow_one, ENNReal.coe_rpow_of_nonneg _ nonneg, coe_ofNat]
```
Is this proof coming from somewhere? If so, can you golf the original using it?

YaelDillies commented on Mathlib/Analysis/Normed/Lp/ProdLp.lean:
```diff
@@ -380,30 +386,30 @@ abbrev prodPseudoMetricAux [PseudoMetricSpace α] [PseudoMetricSpace β] :
+        gcongr
+        simp only [self_le_add_left]
+
+theorem prod_antilipschitzWith_ofLp_aux [PseudoEMetricSpace α] [PseudoEMetricSpace β] :
```
```suggestion
private theorem prod_antilipschitzWith_ofLp_aux [PseudoEMetricSpace α] [PseudoEMetricSpace β] :
```
since it's auxiliary

YaelDillies commented on Mathlib/Analysis/Normed/Lp/ProdLp.lean:
```diff
@@ -563,6 +592,22 @@ theorem prod_infty_equiv_isometry [PseudoEMetricSpace α] [PseudoEMetricSpace β
+    AntilipschitzWith ((2 : ℝ≥0) ^ (1 / p).toReal) (@WithLp.ofLp p (α × β)) :=
+  prod_antilipschitzWith_equiv_aux p α β
+
+theorem prod_infty_ofLp_isometry [PseudoEMetricSpace α] [PseudoEMetricSpace β] :
```
```suggestion
theorem isometry_prod_infty_ofLp [PseudoEMetricSpace α] [PseudoEMetricSpace β] :
```

Paul-Lez replied:
Also changed the original theorem name to `isometry_prod_infty_equiv` for consistency.

YaelDillies replied:
Oh I was thinking you could leave that one alone since you're about to deprecate it anyway

YaelDillies reviewed (approved):
Thanks!

maintainer delegate

eric-wieser commented on Mathlib/Analysis/Normed/Lp/WithLp.lean:
```diff
@@ -50,8 +51,19 @@ variable (p : ℝ≥0∞) (K : Type uK) (K' : Type uK') (V : Type uV)
 
 namespace WithLp
 
+variable {p V} in
```
Should `p` be explicit to one or both of these?

YaelDillies replied:
`p` could be explicit in `toLp`, yes

eric-wieser replied:
That sounds good to me, it's quite common to want to write `‖toLp v‖`, and this won't type-check unless the `p` is specified somehow. The same problem doesn't exist for `ofLp`.

eric-wieser reviewed (commented):
I think we should rip off the bandaid and deprecate `equiv` in _this_ PR, rather than in a follow-up. Otherwise we have two spellings of everything, and reviewers have to tell contributors which one to use.

YaelDillies commented:
Can't this be done in a second PR soon? Paul is really really focused on that refactor and definitely doesn't have 10 other things going on.

eric-wieser commented:
I don't see any benefit in merging this PR until the other is ready to go. Mathlib shouldn't have two non-deprecated ways to spell something identical.

Paul-Lez commented:
I can take care of this tonight!

YaelDillies commented:
Sure, I agree. All I'm saying is that they should be in separate PR. But I am happy for both PRs to land in a short time frame.

Paul-Lez commented:
Added the deprecations (and also removed `simp` tags from deprecated lemmas). 

Along the way, I also noticed that `ofLpProdContinuousLinearEquiv`/`prodContinuousLinearEquiv` should be generalised, and added a TODO note to that effect. I think this may be better to leave to another PR since it would be natural to move it to the `WithLp` file, but the imports aren't quite right (and maybe we don't want to be changing those in this PR?)

eric-wieser commented on Mathlib/Analysis/Normed/Lp/WithLp.lean:
```diff
@@ -85,62 +97,108 @@ instance instModuleFinite
+
+section equiv
+
+@[deprecated "WithLp.equiv has been deprecated, use `WithLp.ofLp` instead" (since := "2024-04-27")]
```
```suggestion
@[deprecated ofLp_zero (since := "2024-04-27")]
```
etc

eric-wieser commented on Mathlib/Analysis/Normed/Lp/WithLp.lean:
```diff
@@ -50,8 +51,19 @@ variable (p : ℝ≥0∞) (K : Type uK) (K' : Type uK') (V : Type uV)
 /-- The canonical equivalence between `WithLp p V` and `V`. This should always be used to convert
 back and forth between the representations. -/
+@[deprecated WithLp.ofLp (since := "2024-04-27")]
 protected def equiv : WithLp p V ≃ V := Equiv.refl _
```
Making this an `abbrev` (but not an `alias`) for `ofLp` should mean that the new `ofLp` lemmas continue to apply to it

eric-wieser commented on Mathlib/Analysis/Normed/Lp/WithLp.lean:
```diff
@@ -85,62 +97,108 @@ instance instModuleFinite
-  { LinearEquiv.refl _ _ with
-    toFun := WithLp.equiv _ _
-    invFun := (WithLp.equiv _ _).symm }
+def ofLpLinearEquiv [Semiring K] [AddCommGroup V] [Module K V] : WithLp p V ≃ₗ[K] V  where
+  __ := LinearEquiv.refl _ _
+  toFun := WithLp.ofLp
+  invFun := WithLp.toLp p
+
+/-- `WithLp.equiv` as a linear equivalence. -/
+@[deprecated WithLp.ofLpLinearEquiv  (since := "2024-04-27")]
+protected def linearEquiv [Semiring K] [AddCommGroup V] [Module K V] : WithLp p V ≃ₗ[K] V where
+  __ := LinearEquiv.refl _ _
+  toFun := WithLp.equiv _ _
+  invFun := (WithLp.equiv _ _).symm
```
There's no need to have both of these; a `deprecated alias` of `linearEquiv` for `ofLpLinearEquiv` would be enough.

leanprover-community-bot-assistant commented:
This pull request has conflicts, please merge `master` and resolve them.

YaelDillies commented on Mathlib/Analysis/InnerProductSpace/PiL2.lean:
```diff
@@ -185,9 +185,19 @@ theorem finrank_euclideanSpace_fin {n : ℕ} :
+    ⟪x, y⟫ = dotProduct (WithLp.ofLp y) (star <| WithLp.ofLp x) :=
+  rfl
+
+theorem EuclideanSpace.inner_piLp_toLp (x y : ι → 𝕜) :
```
```suggestion
theorem EuclideanSpace.inner_toLp_toLp (x y : ι → 𝕜) :
```

YaelDillies commented on Mathlib/Analysis/InnerProductSpace/PiL2.lean:
```diff
@@ -229,7 +239,7 @@ variable (ι 𝕜)
 
 /-- A shorthand for `PiLp.continuousLinearEquiv`. -/
```
```suggestion
/-- A shorthand for `PiLp.ofLpContinuousLinearEquiv`. -/
```

YaelDillies commented on Mathlib/Analysis/InnerProductSpace/PiL2.lean:
```diff
@@ -247,22 +257,32 @@ variable [DecidableEq ι]
+  WithLp.toLp _ (Pi.single i a)
 
 @[simp]
+theorem WithLp.ofLp_single (i : ι) (a : 𝕜) :
```
```suggestion
theorem EuclideanSpace.ofLp_single (i : ι) (a : 𝕜) :
```
for disambiguation

YaelDillies commented on Mathlib/Analysis/InnerProductSpace/PiL2.lean:
```diff
@@ -247,22 +257,32 @@ variable [DecidableEq ι]
   rfl
 
 @[simp]
+theorem WithLp.toLp_single (i : ι) (a : 𝕜) :
```
```suggestion
theorem EuclideanSpace.toLp_single (i : ι) (a : 𝕜) :
```

YaelDillies commented on Mathlib/Analysis/InnerProductSpace/PiL2.lean:
```diff
@@ -1025,30 +1045,56 @@ variable [Fintype n] [DecidableEq n]
+      (WithLp.ofLpLinearEquiv _ 𝕜 (m → 𝕜)).symm
 
 @[simp]
+theorem toEuclideanLin_piLp_toLp (A : Matrix m n 𝕜) (x : n → 𝕜) :
```
```suggestion
theorem toEuclideanLin_toLp (A : Matrix m n 𝕜) (x : n → 𝕜) :
```

YaelDillies commented on Mathlib/Analysis/InnerProductSpace/PiL2.lean:
```diff
@@ -1025,30 +1045,56 @@ variable [Fintype n] [DecidableEq n]
   rfl
 
 @[simp]
+theorem piLp_ofLp_toEuclideanLin (A : Matrix m n 𝕜) (x : EuclideanSpace 𝕜 n) :
```
```suggestion
theorem ofLp_toEuclideanLin (A : Matrix m n 𝕜) (x : EuclideanSpace 𝕜 n) :
```

YaelDillies commented on Mathlib/Analysis/InnerProductSpace/PiL2.lean:
```diff
@@ -1025,30 +1045,56 @@ variable [Fintype n] [DecidableEq n]
   rfl
 
 @[simp]
+theorem piLp_ofLp_toEuclideanLin_apply (M : Matrix m n 𝕜) (v : EuclideanSpace 𝕜 n) :
```
```suggestion
theorem ofLp_toEuclideanLin_apply (M : Matrix m n 𝕜) (v : EuclideanSpace 𝕜 n) :
```

YaelDillies commented on Mathlib/Analysis/InnerProductSpace/PiL2.lean:
```diff
@@ -1067,18 +1113,33 @@ lemma toEuclideanLin_eq_toLin_orthonormal [Fintype m] :
 theorem inner_matrix_row_row [Fintype n] (A B : Matrix m n 𝕜) (i j : m) :
     ⟪A i, B j⟫ₑ = (B * Aᴴ) j i := by
   rfl
 
+/-- The inner product of a row of `A` and a row of `B` is an entry of `B * Aᴴ`. -/
+@[deprecated inner_matrix_row_row (since := "2024-04-27")]
+theorem inner_matrix_row_row' [Fintype n] (A B : Matrix m n 𝕜) (i j : m) :
+    ⟪A i, B j⟫ₑ' = (B * Aᴴ) j i := by
+  rfl
```
I don't see the point of adding a new lemma that's immediately deprecated. You are changing the lemma statement of `inner_matrix_row_row`, so whatever downstream uses of `inner_matrix_row_row` there are, creating a new lemma `inner_matrix_row_row'` won't save the people from having to figure out what happened.
```suggestion
```
Same below

leanprover-community-bot-assistant commented:
This pull request has conflicts, please merge `master` and resolve them.

leanprover-community-bot-assistant commented:
This pull request has conflicts, please merge `master` and resolve them.

leanprover-community-bot-assistant commented:
This pull request has conflicts, please merge `master` and resolve them.

leanprover-community-bot-assistant commented:
This pull request has conflicts, please merge `master` and resolve them.

EtienneC30 commented on Mathlib/Analysis/Normed/Lp/PiLp.lean:
```diff
@@ -581,17 +629,35 @@ theorem nnnorm_eq_ciSup (f : PiLp ∞ β) : ‖f‖₊ = ⨆ i, ‖f i‖₊ :=
+  dsimp only [ofLp_apply]
+
+set_option linter.deprecated false in
+@[deprecated lipschitzWith_ofLp (since := "2024-04-27")]
```
```suggestion
@[deprecated nnnorm_ofLp (since := "2024-04-27")]
```

EtienneC30 commented on Mathlib/Analysis/Normed/Lp/PiLp.lean:
```diff
@@ -581,17 +629,35 @@ theorem nnnorm_eq_ciSup (f : PiLp ∞ β) : ‖f‖₊ = ⨆ i, ‖f i‖₊ :=
+@[simp] lemma nnnorm_toLp (f : ∀ i, β i) : ‖toLp ∞ f‖₊ = ‖f‖₊ := (nnnorm_ofLp _).symm
+
+set_option linter.deprecated false in
+@[deprecated nnnorm_equiv_symm (since := "2024-04-27")]
```
```suggestion
@[deprecated nnnorm_toLp (since := "2024-04-27")]
```

EtienneC30 commented on Mathlib/Analysis/Normed/Lp/WithLp.lean:
```diff
@@ -95,70 +106,117 @@ instance instModuleFinite
-  { LinearEquiv.refl _ _ with
-    toFun := WithLp.equiv _ _
-    invFun := (WithLp.equiv _ _).symm }
+def ofLpLinearEquiv [Semiring K] [AddCommGroup V] [Module K V] : WithLp p V ≃ₗ[K] V  where
```
Would we want `toLpLinearEquiv` as well? I noticed it was in an earlier version, but I didn't see in the discussion why it had been removed.

EtienneC30 commented on Mathlib/MeasureTheory/Measure/Haar/OfBasis.lean:
```diff
@@ -327,13 +327,22 @@ instance [Finite ι] : BorelSpace (EuclideanSpace ℝ ι) := Pi.borelSpace
 @[simps toEquiv]
-protected def measurableEquiv : EuclideanSpace ℝ ι ≃ᵐ (ι → ℝ) where
-  toEquiv := WithLp.equiv _ _
+def measurableOfLp : EuclideanSpace ℝ ι ≃ᵐ (ι → ℝ) where
```
I think this one should rather be named `ofLpMeasurableEquiv`, on the same model as `ofLpLinearEquiv`.

eric-wieser commented:
> Ok, but why would we make `ofLp` and `toLp` plain functions?

Because this permits an eventual future of
```lean-4
structure WithLp where toLp :: ofLp : V
```

leanprover-community-bot-assistant commented:
This pull request has conflicts, please merge `master` and resolve them.
